### PR TITLE
add blacklist

### PR DESCRIPTION
**Pull request by Roger-luo**
as described in #5 I think this at least will solve some for loop slow down due to type inference failure in `iterate` etc. But I haven't investigated much in this direction.

### CODE REVIEW

File: src/replaying.jl [hunk 1]
```diff
@@ -42,6 +41,22 @@ end
 end
 
 
+const WHITE_LIST = [
```
Comment (oxinabox) [inline src/replaying.jl]: I would call this a blacklist.
Its things we don't record

File: src/replaying.jl [hunk 2]
```diff
@@ -42,6 +41,22 @@ end
...
+    Base.promote_op, Base.to_shape,
+    Core.getfield,
+    Core.:(===),
+    Base.iterate,
+    Broadcast.broadcasted, Broadcast.instantiate,
```
Comment (oxinabox) [inline src/replaying.jl]: won't blocking `Broadcast.instantiate` cause a ton of allocations to be missed?

Comment (Roger-luo) [reply src/replaying.jl]: but the allocation happens in `materialize!` which is the `copyto!` function IIUC?

**Comment by oxinabox** [comment]
Are tests passing for you locally?
They won't run on CI without:
https://github.com/JuliaCI/BenchmarkTools.jl/issues/158

**Comment by Roger-luo** [comment]
I see. I didn't notice you are using that for allocation, I was just using `@allocated` to test allocations.

I got this locally, but I think this is actually an improvement ;-)

```julia
matmul example: Test Failed at /Users/roger/.julia/dev/AutoPreallocation/test/integration_tests.jl:19
  Expression: #= /Users/roger/.julia/dev/AutoPreallocation/test/integration_tests.jl:19 =# @ballocated(avoid_allocations($(Expr(:$, :record)), f_matmul)) == 352
   Evaluated: 256 == 352
Stacktrace:
```

I changed the test allocation, should pass locally now.

**Comment by Roger-luo** [comment]
Oh, BTW, could you register the package for v0.1 since it's announced already anyways? I'm switching to it for my own project from my old implementation.

And one question, what's the difference between `Base.Typeof` (which is used in Cassette) and `typeof`? I find if I define these "backlist" functions with `Base.Typeof` there will be a huge slow down. I don't understand why however.

**Comment by oxinabox** [comment]
I will tag release soon

> And one question, what's the difference between Base.Typeof (which is used in Cassette) and typeof? I find if I define these "backlist" functions with Base.Typeof there will be a huge slow down. I don't understand why however.

I don't know. I have been wondering also

**Comment by oxinabox** [comment]
It should be possible to make test passed now because a release of NenchmarkTools has been tagged.

Need up the lower bounds in compat

**Comment by Roger-luo** [comment]
I don't think CI fails because of this PR. But it seems 1.0 has different allocation comparing to 1.3 somehow...

**Comment by Roger-luo** [comment]
Actually I think we should black list "instanitiate" at least in the following case:

```
x = rand(1000)
bc = Broadcast.broadcasted(tanh, x)
_, record = record_allocations(Broadcast.materialize, bc)
ctx = AutoPreallocation.new_replay_ctx(record)
Cassette.overdub(ctx, Broadcast.materialize, bc)

@benchmark for _ in 1:1000_00
    Broadcast.instantiate(bc)
end

@benchmark for _ in 1:1000_00
    Cassette.overdub(AutoPreallocation.new_replay_ctx(record), Broadcast.instantiate, bc)
end
```

Preallocation version allocates more memory and is 3x slower.

File: .travis.yml [hunk 3]
```diff
@@ -5,6 +5,7 @@ os:
   - osx
 julia:
   - 1.0
```
Comment (oxinabox) [inline .travis.yml]: Since tracking exact allocations is hard lets not test on 1.0
```suggestion
```

File: test/integration_tests.jl [hunk 4]
```diff
@@ -16,5 +16,5 @@ end
 @testset "matmul example" begin
     @assert (@ballocated f_matmul()) === 18_304
     val, record = record_allocations(f_matmul)
-    @test (@ballocated avoid_allocations($record, f_matmul)) == 352
+    @test (@ballocated avoid_allocations($record, f_matmul)) == 256
```
Comment (oxinabox) [inline test/integration_tests.jl]: This line is failing on CI says 320 allocations?

Comment (Roger-luo) [reply test/integration_tests.jl]: errr... I got 256 on my machine however. 

```julia
julia> versioninfo()
Julia Version 1.3.0
Commit 46ce4d7933 (2019-11-26 06:09 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.0.0)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

what about changing it to <=252 then?

**Comment by Roger-luo** [comment]
@oxinabox seems good?

**Comment by oxinabox** [comment]
thanks

**Pull request merged**